File: websocket-prototype/java-processor/app/javascript/channels/java_runner_channel.js
import consumer from "./consumer"

const helloWorld = "public class HelloWorld {public static void main(String[] args) {"
+
  "\t\tSystem.out.println(\"Hello World\");\n"
+
  "\t}\n"
+
  "}\n"

document.addEventListener('turbolinks:load', () => {

  const userId = document.getElementById('user-id').getAttribute('data-user-id');
  let connected = false;

  const connection = consumer.subscriptions.create({channel: "JavaRunnerChannel", user_id: userId}, {
    connected() {
      console.log("connected to channel " + userId);
      connected = true;
    },

    disconnected() {
      // Called when the subscription has been terminated by the server
    },

    received(data) {
      console.log(data);
      const outputDiv = document.getElementById('program-output')
      console.log(outputDiv);
      const stdout = data.stdout
      const stderr = data.stderr
      const status = data.status
      const html = "stdout: " + stdout + " stderr: " + stderr + " status: " + status + "\n"
      outputDiv.innerHtml = outputDiv.innerHtml + html
    },

    run_program(filename, text) {
      console.log("calling run program");
      this.perform("run_program", {filename: filename, text: text, user_id: userId})
    }
  });
  console.log(connection);
  connection.run_program("HelloWorld.java", helloWorld);
})

const sleep = (milliseconds) => {
  return new Promise(resolve => setTimeout(resolve, milliseconds))
}

File: websocket-prototype/java-processor/app/channels/java_runner_channel.rb
class JavaRunnerChannel < ApplicationCable::Channel
  @@java_runners = {}

  def subscribed
    # stream_from "some_channel"
    stream_from "java_runner_channel_#{params[:user_id]}"
  end

  def unsubscribed
    # Any cleanup needed when channel is unsubscribed
  end

  def run_program(data)
    user_id = data["user_id"]
    runner = @@java_runners[user_id]
    if (!runner)
      runner = JavaRunner.new(user_id)
      @@java_runners[user_id] = runner
    end
    if (runner.currently_running)

      ActionCable.server.broadcast "java_runner_channel_#{user_id}", "This user is already running a program"
    end
    runner.run_java_program(data["filename"], data["text"])
  end

  def send_input(data)
    user_id = data["user_id"]
    runner = @@java_runners[user_id]
    if !runner
      ActionCable.server.broadcast "java_runner_channel_#{user_id}", "This user is already running a program"
    end
    runner.addToStdin(data["data"])
  end
end
